scipy/sparse/linalg/dsolve/tests/test_linsolve.py

```diff
@@ -498,5 +498,52 @@ def worker():
         assert_equal(len(oks), 20)
 
 
+class TestSpsolveTriangular(TestCase):
+
+    def test_singular(self):
+        n = 5
+        A = csr_matrix((n,n))
+        b = np.arange(n)
+        for lower in (True, False):
+            assert_raises(scipy.linalg.LinAlgError, spsolve_triangular, A, b, lower=lower)
+
+    def test_bad_shape(self):
+        # A is not square.
+        A = np.zeros((3, 4))
+        b = ones((4, 1))
+        assert_raises(ValueError, spsolve_triangular, A, b)
+        # A2 and b2 have incompatible shapes.
+        A2 = csr_matrix(eye(3))
+        b2 = array([1.0, 2.0])
+        assert_raises(ValueError, spsolve_triangular, A2, b2)
+
+    def test_input_types(self):
+        A = array([[1., 0.], [1., 2.]])
+        b = array([[2., 0.], [2., 2.]])
+        for matrix_type in (array, csc_matrix, csr_matrix):
+            x = spsolve_triangular(matrix_type(A), b, lower=True)
+            assert_array_almost_equal(np.dot(A, x), b)
```

Comment:
Using `A.dot(x)` here is better, because it calls the sparse matrix `dot` directly instead of routing through numpy.